src/Numerics/OptimizableFunctorBase.h

```diff
@@ -113,5 +123,22 @@ struct OptimizableFunctorBase
 
 };
 
+inline void print(OptimizableFunctorBase& func, std::ostream& os)
```

Comment:
A free function in the header may cause multi-definition problem. Make it a template(functor) function.

Reply:
I'm not sure what you mean by a template(functor) function.  Do you want it to become a method of OptimizableFunctorBase?  Regarding the multi-definition problem, my understanding was that making it inline solved this issue.  I don't think the function is generic beyond this class hierarchy so I'm not sure that doing something like:
template<class foo>
void print(foo& f, std::ostream& os) 
makes sense to me.  Is that what you had in mind?

Reply:
Yes.
```C++
template<class FT>
void print(const FT& f, std::ostream& os) 
```

Reply:
If the compilation becomes very slow, we may break one cpp into several and then we may encounter the multi-definition problem. inline is a weak suggestion to the compiler so not necessarily followed by the compiler.

Reply:
I'm a bit uneasy with the template suggestion.  To me that implies that the function is more general than it is (it's only meant to apply to OptimizableFunctorBase derived objects).  In terms of capturing my intent, I would actually be happier leaving the declaration in the header and putting the definition in a cpp file.  That said, I'll defer to your judgement.  Should I change it to a template?

Reply:
I don't insist on template. Add a cpp file is also fine to me although there is one more file and you need to edit the cmake file to include it.

Reply:
This is a simple "service" function with no performance requirements. The inline is definitely not needed here (as with many uses of inline in our source code). Why not put this function in the .cpp file? Ideally it would be a member function, but provided it is legal just keep it simple.

Luke won't be able to clean up all this code so it is important to strike a balance.

Reply:
OK, making the switch now.  One question though.  Your snippet made the argument const.  That was my first inclination, but the f and df methods of OptimizableFunctorBase are not declared const even though it seems to me they could be.  Should I go ahead and change them and track down the derived classes?

Reply:
If it is not too painful due to the cascading dependencies, improving const correctness is worthwhile.

Reply:
@lshulen if f and df are not const, you can not put the const. Just leave it.

Reply:
I'm trying to make f and df const because I think this is the right thing to do.  Mostly it's not causing a problem.  However, it seems that CubicBSpline is a state machine, so while the intention is that nothing is being changed by calls like splint, there are internal scratch variables that are being changed.  

I'm going to take a bit harder look at this and see if I can make it work out.  Otherwise, I'll just make the argument to print non-const and go from there.

Reply:
I believe making CubicBspline non-const will be quite a lot of work. We can have the change on builders merged in and then solve the non-const problem. I don't quite understand why we have this one and OneDimCubicSpline which seems doing similar things.

Reply:
I just pushed a change which makes the current code base work at the cost of making a few scratch variables mutable.  To me it made sense to do this and then make an issue to get rid of CubicBspline in favor of OneDimCubicSpline or a direct use of a one dimensional bspline.  I'll be curious to see if you are of the same opinion.

Reply:
IMO, marking scratch space mutable is dangerous and misleading. When I read the code and see const function. I would expect the code is thread safe and I can call f in a threaded region. With a mutable scratch this is not true. If the purpose is only to get the code to run, I'd like to leave the code without const so it doesn't confuse people. Mutable can be used for some internal parameter which may need some change after calling the constructor but having frequently changing data mutable lost of validity of const.

Reply:
OK, I'll unwind the const business.  In the end, I find myself not happy with the architecture of these sections.  As @ye-luo rightly points out, they are not thread-safe and I can't see how to make them so without serious redesign.  Also, the places I can find them used in the code don't appear to derive much benefit from the scratch spaces other than perhaps avoiding allocation of a few bytes of memory on call.  I was sort of expecting staged evaluation where the bits left in the scratch space were re-used but I don't see where this is happening.

I think I will make an issue and come back to the general design of the functors later.